### PR TITLE
gh-102153: Start stripping C0 control and space chars in `urlsplit`

### PR DESCRIPTION
**Pull request by illia-v**
`urllib.parse.urlsplit` has already been respecting the WHATWG spec a bit #25595.

This adds more sanitizing to respect the "Remove any leading C0 control or space from input" [rule](https://url.spec.whatwg.org/#url-parsing:~:text=Remove%20any%20leading%20and%20trailing%20C0%20control%20or%20space%20from%20input.) in response to [CVE-2023-24329](https://nvd.nist.gov/vuln/detail/CVE-2023-24329).

* Issue: gh-102153

### CODE REVIEW

**Comment by xiaoge1001** [comment]
It's great. it handle some bad input URLs and make urlsplit more robust.

https://url.spec.whatwg.org/#concept-basic-url-parser

**Comment by xiaoge1001** [comment]
Does anyone review https://github.com/python/cpython/pull/102470 or https://github.com/python/cpython/pull/102508?

**Comment by gpshead** [comment]
This PR moves us in the direction I'd like to see the APIs go.  I'm not sure it could ever be backported as a security fix for the long list of analysis reasons I just posted over in the Issue, but it might make sense in 3.12.

I believe there's still an open question of trailing spaces on paths potentially carrying meaning in some (presumably not-HTTP) scenarios (see the Issue analysis). So stripping all of the C0 characters early on might not be the right approach. But they should be stripped from most places.  Lets keep this PR around as a likely option.

(I'll still put a do-not-submit tag on it for now, just as an indicator that we're not ready to accept it _yet_).

**Comment by xiaoge1001** [comment]
Is it planned to merge the current patch to python3.12? I don't think there's been any discussion recently.

**Comment by gpshead** [comment]
Results from internal testing across a huge body of code (including test suites of many open source libraries and applications): Stripping the C0 characters from the left hand side doesn't appear to cause any problems.  Stripping them from the right hand side of the URL does, as existing code calling urllib.parse such as `django.core.validators.URLValidator.__call__` (at least up through 3.2, the code has changed slightly in 4.2 but I believe that remains true).

i've updated the PR to use lstrip on the url.

**Comment by gpshead** [comment]
I'm a lot more confident that we could actually backport this as a security bug fix without breaking the world now.

As for documentation to be added to warn users of these APIs, I'm not quite sure how to word that and what to convey yet. I've opened up as a discussion in https://discuss.python.org/t/how-to-word-a-warning-about-security-uses-in-urllib-parse-docs/26399.

File: Doc/library/urllib.parse.rst [hunk 1]
```diff
@@ -324,8 +324,9 @@ or on combining URL components into a URL string.
    ``#``, ``@``, or ``:`` will raise a :exc:`ValueError`. If the URL is
    decomposed before parsing, no error will be raised.
 
-   Following the `WHATWG spec`_ that updates RFC 3986, ASCII newline
-   ``\n``, ``\r`` and tab ``\t`` characters are stripped from the URL.
+   Following some of the `WHATWG spec`_ that updates RFC 3986, leading C0
+   control control and space characters are stripped from the URL. ``\n``,
```
Comment (frenzymadness) [inline Doc/library/urllib.parse.rst]:
```suggestion
   control and space characters are stripped from the URL. ``\n``,
```

File: Misc/NEWS.d/next/Security/2023-03-07-20-59-17.gh-issue-102153.14CLSZ.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+:func:`urllib.parse.urlsplit` now strips leading and trailing C0 control and
```
Comment (frenzymadness) [inline Misc/NEWS.d/next/Security/2023-03-07-20-59-17.gh-issue-102153.14CLSZ.rst]: Is the "trailing" correct here?

**Review by frenzymadness — COMMENTED** [review]
The first beta is expected to be released on Monday and it'd be very nice to have this in the release.

File: Misc/NEWS.d/next/Security/2023-03-07-20-59-17.gh-issue-102153.14CLSZ.rst (hunk 2 above)
Comment (gpshead) [reply Misc/NEWS.d/next/Security/2023-03-07-20-59-17.gh-issue-102153.14CLSZ.rst]: good catch, i thought I had found all of those references. fixed.

**Comment by gpshead** [comment]
I intend to get this in before the next releases are cut. The last bit is for me to add some appropriate text to the urllib.parse documentation recommending that people verify that the outputs from the API makes sense before trusting any of them.

**Comment by frenzymadness** [comment]
We have one week to the first beta. Is there anything I can help you with?

**Comment by gpshead** [comment]
> We have one week to the first beta. Is there anything I can help you with?

I was waiting until I had (re)written some text to add to the urllib.parse docs after the discuss thread linked above.  I've pushed a documentation update in a commit.  Take a look.

I believe this is ready to go in and is something we can safely backport to older releases.  I'll deal with the backports manually one by one as they're each likely to need some fixups.

The documentation may need further tweaks - but that can always be iterated on later as deemed appropriate.

**Review by gpshead — APPROVED** [review]

**Comment by illia-v** [comment]
@gpshead thank you for making the extensive research and needed changes!

**Pull request merged**